An even better improvement to the dedupe() section

amontalenti opened the pull request:
This takes the "best" rewrite of `dedupe()` from PR #19, but doesn't go overboard with a belabored discussion.

This PR also improves the declarative vs imperative section to link to Python's "Functional Programming HOWTO", which is an official part of the docs. So, this also addresses issues #10 and #21.

Would love to hear what these folks have to say about the changes:
- @TheNeuralBit (who asked for more details on declarative vs imperative)
- @j6k4m8 (who raised the issue of `dedupe()` being confusing in current version)
- @veprbl (who pointed out my first attempt at expanding this section was convoluted)

Hopefully this PR addresses all concerns.

amontalenti committed:
An even better improvement to the dedupe() section

This takes the "best" rewrite from another branch, but doesn't introduce
a new rule or too much verbosity in the example.

amontalenti committed:
An even better improvement to the dedupe() section

This takes the "best" rewrite from another branch, but doesn't introduce
a new rule or too much verbosity in the example.

TheNeuralBit commented:
I still like the idea of adding a concise inline explanation of the difference between declarative and imperative as well, but just providing a link is a step in the right direction

veprbl commented:
You still write:

```
As a simple example, you should **never** write code like this:
```

And I still would totally write a `dedup` like this in certain case when I don't need ordering. It would be without return statement or with making a full copy of the original list and without premature optimisation of reversing indices (btw. you don't need to "sort" them, just reverse: `dupe_positions[::-1]`). Because python (at least in versions that I use — 2.6, 2.7) just doesn't have anything like unordered_set there is no king's way around this as far as I know. The original dedup with a slight modification can be a perfectly fine function and can be used in production.

If you want "fast" inplace version:

``` python
# perfectly fine
def dedupe(items):
    seen = set()
    dupe_positions = []
    for i, item in enumerate(items):
        if item in seen:
            dupe_positions.append(i)
        else:
            seen.add(item)
    for idx in dupe_positions[::-1]:
        items.pop(idx)
```

If you want num_dupes:

``` python
# perfectly fine too
def dedupe(items):
    result = []
    num_dupes = 0
    seen = set()
    for item in items:
        if item not in seen:
            result.append(item)
            seen.add(item)
        else:
            num_dupes += 1
    return result, num_dupes
```

But you claim that this should **never** happen.

amontalenti commented:
@veprbl What if I said you should "avoid" code like this, rather than saying you should never write code like that.

I think you are getting lost in the "ordering" detail which is irrelevant to this discussion (of pure functions). Both of your "fine" rewrites mutate `items` in-place.

Also, I'll change it to use `reversed()`, which is more idiomatic than the original `sorted(..., reverse=True)` and the negative stride slice syntax you suggested. You should also know that the slice syntax is eager and forces a copy of the indices list.

amontalenti committed:
Better dedupe() and declarative description

Addresses concerns raised by @TheNeuralBit and @veprbl in
PR #23.

amontalenti committed:
Merge branches

veprbl commented:
> On 06 Jan 2016, at 09:08, Andrew Montalenti notifications@github.com wrote:
> 
> @veprbl What if I said you should "avoid" code like this, rather than saying you should never write code like that.
> But it says "you should **never** write code like this”.
> 
> I think you are getting lost in the "ordering" detail which is irrelevant to this discussion (of pure functions).
> It’s a detail, but you are breaking a contract where good and bad version are interchangeable. They should at least have the same return type!
> Both of your "fine" rewrites mutate items in-place.
> Second doesn't
> 
> Also, I'll change it to use reversed(), which is more idiomatic than the original sorted(..., reverse=True) and the negative stride slice syntax you suggested. You should also know that the slice syntax is eager and forces a copy of the indices list.
> 
> —
> Reply to this email directly or view it on GitHub.

veprbl commented:
> On 06 Jan 2016, at 09:08, Andrew Montalenti notifications@github.com wrote:
> 
> @veprbl What if I said you should "avoid" code like this, rather than saying you should never write code like that.
> Sorry misread you in last email. Yes, that would be better.
> 
> I think you are getting lost in the "ordering" detail which is irrelevant to this discussion (of pure functions). Both of your "fine" rewrites mutate items in-place.
> 
> Also, I'll change it to use reversed(), which is more idiomatic than the original sorted(..., reverse=True) and the negative stride slice syntax you suggested. You should also know that the slice syntax is eager and forces a copy of the indices list.
> 
> —
> Reply to this email directly or view it on GitHub.

amontalenti commented:
@veprbl Check out the latest diff. It addresses all your concerns. I changed the language, switched to `reversed`, and the return types are matching.

@TheNeuralBit I also added a sentence derived from your contribution that is a quick, in-line description of declarative vs imperative.

Good to go? :question: :+1:

TheNeuralBit commented:
Good to go :+1:

veprbl commented:
> On 06 Jan 2016, at 09:18, Andrew Montalenti notifications@github.com wrote:
> return types are matching.
> set != list

amontalenti commented:
@veprbl it's a 2-tuple of `(deduped, num_removed)`, whether `deduped` is a list, set, or deque is an implementation detail, IMO.

If it helps, you could think of it as `isinstance(deduped, abc.Iterable) is True` is the contract offered by the function. This is common in many places in Python so it doesn't worry me. I shall merge since your other concerns are addressed.

(ed: made some edits to this comment)
